lib/layer.cpp

```diff
@@ -0,0 +1,447 @@
+
+const _cl_icd_dispatch &layer::get_target_dispatch() const {
+  return *tdispatch;
+}
+
+bool layer::device_supports_spirv_via_extension(cl_device_id device,
+                                                cl_int *err) {
+  // If querying extensions, check if SPIR-V is supported
+  size_t dispatch_param_value_size_ret = 0;
+  cl_int err_ = tdispatch->clGetDeviceInfo(
+      device, CL_DEVICE_EXTENSIONS, 0, nullptr, &dispatch_param_value_size_ret);
+  if (err_ == CL_SUCCESS) {
+    std::vector<char> dispatch_extensions(dispatch_param_value_size_ret, '\0');
+    err_ = tdispatch->clGetDeviceInfo(device, CL_DEVICE_EXTENSIONS,
+                                      dispatch_extensions.size(),
+                                      dispatch_extensions.data(), nullptr);
+    if (err_ == CL_SUCCESS) {
+      std::string_view dispatch_extensions_string{dispatch_extensions.data(),
+                                                  dispatch_extensions.size()};
+      static const char *spirv_ext_name = "cl_khr_il_program";
+      if (dispatch_extensions_string.find(spirv_ext_name) != std::string::npos)
+        return true;
+      else
+        return false;
+    } else {
+      if (err != nullptr)
+        *err = err_;
+      return false;
+    }
+  } else {
+    if (err != nullptr)
+      *err = err_;
+    return false;
+  }
+}
+
+bool layer::spirv_queries_are_core_for_device(cl_device_id device,
+                                              cl_int *err) {
+  cl_platform_id device_platform;
+  cl_int err_ = tdispatch->clGetDeviceInfo(device, CL_DEVICE_PLATFORM,
+                                           sizeof(cl_platform_id),
+                                           &device_platform, nullptr);
+  if (err_ == CL_SUCCESS) {
+    size_t platform_version_size_ret = 0;
+    err_ = err_ =
+        tdispatch->clGetPlatformInfo(device_platform, CL_PLATFORM_VERSION, 0,
+                                     nullptr, &platform_version_size_ret);
+    if (err_ == CL_SUCCESS) {
+      std::string platform_version(platform_version_size_ret, '\0');
+      err_ = tdispatch->clGetPlatformInfo(device_platform, CL_PLATFORM_VERSION,
+                                          platform_version.size(),
+                                          platform_version.data(), nullptr);
+      if (err_ == CL_SUCCESS) {
+        if (platform_version.find("OpenCL 2.1") != std::string::npos ||
+            platform_version.find("OpenCL 2.2") != std::string::npos) {
+          if (err != nullptr)
+            *err = CL_SUCCESS;
+          return true;
+        } else {
+          if (err != nullptr)
+            *err = CL_SUCCESS;
+          return false;
+        }
+      } else {
+        if (err != nullptr)
+          *err = CL_SUCCESS;
+        return false;
+      }
+    } else {
+      if (err != nullptr)
+        *err = CL_SUCCESS;
+      return true;
+    }
+  } else {
+    if (err != nullptr)
+      *err = CL_SUCCESS;
+    return false;
+  }
```

Comment:
I'm not too happy with the deep nesting of code as well. Personally I'm in favor of making the "happy path" of code as straightforward as possible, and error handling be centrallized, or in some other way "out of sight" not to increase noise. I'm not a big fan of the [goto-style error handling](https://github.com/KhronosGroup/OpenCL-SDK/blob/main/samples/core/reduce/main.c#L519-L539) what the C samples exhibit, but it does tick all the boxes for me. (It's manual exception handling, really.) I was thinking of using the C++ API here, but that wraps the global namespace API and not the dispatch table. This use case mildly resembles [this](https://github.com/KhronosGroup/OpenCL-CLHPP/pull/170) issue (that refers to the wrapping name, not the wrapped contents), but we may be able to hit two birds with one stone.

Reply:
I'm not advocating for goto and I don't think you'd need goto's here. I think early returns would work well:
```cpp
err = call();
if (err != SUCCESS) {
  return err;
}

err = call2();
...
```

Granted that means the "happy path" is cluttered with error checking code but I think it would still be easier to read.

Reply:
I put some early returns in the code to make it easier to read.